src/transaction/log_page_buffer.c

```diff
@@ -2203,8 +2193,118 @@ logpb_respond_fetch_log_page_request (THREAD_ENTRY &thread_r, std::string &paylo
       // pack page data too
       payload_in_out.append (reinterpret_cast<const char *> (lr.get_page ()), LOG_PAGESIZE);
     }
+  // *INDENT-ON*
+}
+
+static int
+logpb_request_log_hdr_page_from_page_server (LOG_PAGE * log_pgptr)
+{
+  const bool perform_logging = prm_get_bool_value (PRM_ID_ER_LOG_READ_LOG_PAGE);
+  if (perform_logging)
+    {
+      _er_log_debug (ARG_FILE_LINE, "[READ LOG] Sent request for log to Page Server. Page ID: %lld \n",
+		     LOGPB_HEADER_PAGE_ID);
+    }
+  std::string response_message;
+  int error_code = ts_Gl->send_receive (tran_to_page_request::SEND_LOG_HDR_PAGE_FETCH,
+					"", response_message);
+  // there are two layers of errors to he handled here:
+  //  - client side communication to page server error
+  //  - page server side errors
+
+  // client side communication to page server error
+  if (error_code != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      if (perform_logging)
+	{
+	  _er_log_debug (ARG_FILE_LINE, "[READ LOG] Received error log page message from Page Server. Error code: %d\n",
+			 error_code);
+	}
+
+      // TODO handling the case such as shutdown
+      assert_release (error_code != ER_CONN_NO_PAGE_SERVER_AVAILABLE);
+      return error_code;
+    }
+
+  assert (response_message.size () > 0);
+  const char *message_ptr = response_message.c_str ();
+  std::memcpy (&error_code, message_ptr, sizeof (error_code));
+  message_ptr += sizeof (error_code);
+
+  if (error_code == NO_ERROR)
+    {
+      PGLENGTH log_page_size = 0;
+      std::memcpy (&log_page_size, message_ptr, sizeof (LOG_PAGESIZE));
+      message_ptr += sizeof (LOG_PAGESIZE);
+
+      assert_release (log_page_size >= IO_MIN_PAGE_SIZE
+		      && log_page_size <= IO_MAX_PAGE_SIZE && IS_POWER_OF_2 (log_page_size));
+
+      std::memcpy (log_pgptr, message_ptr, log_page_size);
+      message_ptr += log_page_size;
+
+      assert (LOGPB_HEADER_PAGE_ID == log_pgptr->hdr.logical_pageid);
+
+      if (perform_logging)
+	{
+	  _er_log_debug (ARG_FILE_LINE, "[READ LOG] Received log page message from Page Server. Page ID: %lld\n",
+			 log_pgptr->hdr.logical_pageid);
+	}
+    }
+  else
+    {
+      if (error_code == ER_LOG_PAGE_CORRUPTED)
+	{
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_LOG_PAGE_CORRUPTED, 1, LOGPB_HEADER_PAGE_ID);
+	}
+      else
+	{
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_GENERIC_ERROR, 0);
+	}
+      if (perform_logging)
+	{
+	  _er_log_debug (ARG_FILE_LINE, "[READ LOG] Received error log page message from Page Server. Error code: %d\n",
+			 error_code);
+	}
+    }
+  return error_code;
+}
+
+void
+logpb_respond_fetch_log_hdr_page_request (THREAD_ENTRY & thread_r, std::string & payload_in_out)
```

Comment:
I agree with the consistency remark. I am to "blame" since I wrote the initial `logpb_respond_fetch_log_page_request` sister function.

Question:
I know that the THREAD_ENTRY was added relatively recently - sometime around 2015-16 maybe? - and I want to know if it is safe to assume that, when `SERVER_MODE` is active, the `THREAD_ENTRY` is also assumed to be present to matter what/when/where. I know that there are few functions where there is still this sequence
```
  if (thread_p == NULL)
    {
      thread_p = thread_get_thread_entry_info ();
    }
```
bot only ~80 occurrences among the thousands of functions that cubrid has defined (both module-public and module-private).

Reply:
> bot only ~80 occurrences among the thousands of functions that cubrid has defined

just that; in Romanian, when translated ad-litteram, it makes perfect sense :)

Trying to rephrase:
but there are only ~80 occurrences among the thousands of functions that cubrid codebase is composed of

Reply:
> My understanding is that thread_get_thread_entry_info returns not-null thread entry only when the thread calling it has its own entry from cubthread::manager. That is stored in the thread's local variable. This works only when it's a daemon or worker thread of a worker pool spawned by the cubthread::manager. So, if a thread is created just by pthread or std::thread bypassing the manager, it doesn't have a valid entry. Does this answer you?

Yes. Somehow I should have already know this - after all, it is in the code. :)
So, this leads me to think that using the reference instead of the pointer has no drawbacks in this particular situation.

Side note, the passing of the thread entry as argument down the callstack has performance as its main benefit because retrieving a `thread_local` variable in each place it is needed is much more costly than accessing a stack variable (as it boils down to when passed as an argument).